rasa/core/policies/unexpected_intent_policy.py

```diff
@@ -0,0 +1,914 @@
+        label_quantiles = io_utils.pickle_load(
+            model_path / f"{cls._metadata_filename()}.label_quantiles.pkl"
+        )
+        model_utilties.update({"label_quantiles": label_quantiles})
+        return model_utilties
+
+    @classmethod
+    def _update_loaded_params(cls, meta: Dict[Text, Any]) -> Dict[Text, Any]:
+        meta = train_utils.override_defaults(cls.defaults, meta)
+        return meta
+
+    @classmethod
+    def _load_policy_with_model(
+        cls,
+        model: "IntentTED",
+        featurizer: TrackerFeaturizer,
+        model_utilities: Dict[Text, Any],
+        should_finetune: bool,
+    ) -> "UnexpecTEDIntentPolicy":
+        return cls(
+            featurizer=featurizer,
+            priority=model_utilities["priority"],
+            model=model,
+            fake_features=model_utilities["fake_features"],
+            entity_tag_specs=model_utilities["entity_tag_specs"],
+            should_finetune=should_finetune,
+            label_quantiles=model_utilities["label_quantiles"],
+            **model_utilities["meta"],
+        )
+
+
+class IntentTED(TED):
+    """Follows TED's model architecture from https://arxiv.org/abs/1910.00486.
+
+    However, it has been re-purposed to predict multiple
+    labels (intents) instead of a single label (action).
+    """
+
+    def _prepare_label_classification_layers(self, predictor_attribute: Text) -> None:
+        """Prepares layers & loss for the final label prediction step."""
+        self._prepare_embed_layers(predictor_attribute)
+        self._prepare_embed_layers(LABEL)
+        self._prepare_dot_product_loss(LABEL, self.config[SCALE_LOSS])
+
+    def _prepare_dot_product_loss(
+        self, name: Text, scale_loss: bool, prefix: Text = "loss",
+    ) -> None:
+        self._tf_layers[f"{prefix}.{name}"] = self.dot_product_loss_layer(
+            self.config[NUM_NEG],
+            scale_loss,
+            similarity_type=self.config[SIMILARITY_TYPE],
+        )
+
+    @property
+    def dot_product_loss_layer(self) -> tf.keras.layers.Layer:
+        """Returns the dot-product loss layer to use.
+
+        Multiple intents can be valid simultaneously, so `IntentTED` uses the
+        `MultiLabelDotProductLoss`.
+
+        Returns:
+            The loss layer that is used by `_prepare_dot_product_loss`.
+        """
+        return layers.MultiLabelDotProductLoss
+
+    @staticmethod
+    def _get_labels_embed(
+        label_ids: tf.Tensor, all_labels_embed: tf.Tensor
+    ) -> tf.Tensor:
+        # instead of processing labels again, gather embeddings from
+        # all_labels_embed using label ids
+
+        indices = tf.cast(label_ids[:, :, 0], tf.int32)
+
+        # Find padding indices. They should have a value -1
+        padding_indices = tf.where(tf.equal(indices, -1))
+
+        # Create a tensor of ones which will serve as updates to original `indices`
+        updates_to_indices = tf.ones((tf.shape(padding_indices)[0]), dtype=tf.int32)
```

Comment:
In line with Johannes's comment just above, I think here we should `tf.fill` a tensor with `-1 * LABEL_PAD_ID` instead of using `tf.ones` (which assumes the value of `LABEL_PAD_ID`).

Reply:
@samsucik `-1 * LABEL_PAD_ID` will also assume that `LABEL_PAD_ID` is `-1` or at least a negative number. I don't think we can really decouple the implementation to handle padding without knowing what the exact value of padding. Some ops will need to know the value.

Reply:
@dakshvar22 I disagree. If our sole aim is to get 0s (for `tf.gather` to work correctly), then we can blindly sum up `LABEL_PAD_ID + (-1 * LABEL_PAD_ID)` whatever the value of `LABEL_PAD_ID` and we'll always get 0s, as long as `LABEL_PAD_ID` is numerical...

Reply:
Ahh, I see what you mean now. Hmm, good thought. Let me change that.

Reply:
@samsucik I changed the implementation. Using `tf.fill` wasn't possible because it demands a known shape value whereas the `tf.where` operation would return a `None` for the first dimension of `padding_indices`. I used `tf.ones` but made it generic so that even if `LABEL_PAD_ID` is changed to any other numerical value, it should neutralize all padding indices to `0`.

Reply:
Awesome, thanks @dakshvar22!